Implement IAM support

indiv0 opened the pull request:
# Status

* [x] Add `iam` feature
* [x] Add `iam` module
* [x] Enable service generation for `iam` in `build.rs`
* [x] Modify `timestamp` handling to assume it is an `f64`, not a `String`
* [x] Ensure that shapes have the proper case
    * AWS types for `IAM` shapes are `camelCased`, not `CamelCased` like the rest of the AWS services.
* [x] Fix XML deserialization for `IAM` responses
* [x] Verify serialization was not broken for other services (e.g. SQS)
* [x] Fix the way the generator generates function names
    * @cmsd2 has fixed this [upstream](https://github.com/whatisinternet/inflector/commit/6b5614c721ece302f4e98d4f370563f184672bc3).
    * Currently it attempts to naively snake case the type, which doesn't work for types like `ListSAMLProviders`, which becomes `list_s_a_m_l_providers`
* [x] Add tests
    * [x] `list_users`
    * [x] `get_users`
    * [x] protocol tests (implemented outside of this PR)
* [x] Update `README.md`
* [x] Wait for https://github.com/rusoto/rusoto/pull/386 to land, then rebase the code to utilize it's `capitalize_first_letter_of_string` function instead.
* [x] Remove or make [these](https://github.com/rusoto/rusoto/pull/235/files#diff-ff8d2313f22100cddc2afca5c3e20414R358) lines debug-only to reduce codegen memory usage
* [x] Squash commits
* [x] Rebase on master
* [x] Fix the usage of `MockResponseReader` by EC2
* [ ] Await code-review

# Future Issues

* Fix `timestamp` deserialization to deserialize to `chrono::DateTime<FixedOffset>` (#388)
* Open upstream botocore issue to fix mismatched expected/received `ListInstanceProfiles, ListSigningCertificates, ListSigningCertificatesResult, ListAccessKeys` responses

indiv0 commented:
9f6e3ed fixes the casing issues for shape names by implementing a custom serializer for `String` and `BTreeMap` and then using it anywhere a shape name needs to be deserialized.

This is necessary because `IAM` shape names use `camelCase` and not `CamelCase`.
Additionally, I chose not to use `Inflector`'s `to_class_case` function for this because it's incredibly slow (IIRC it iterates over every character in the string) and it actually slows compile speed to a crawl.
I added a custom function which simply uppercases the first letter of the shape name string, which is `O(1)`. However, my function doesn't handle multi-code-point characters, but this should be fine since the AWS service definitions appear to be entirely ASCII.

If this solution is acceptable to you, let me know. If not, I'm open to suggestions on how to fix it.

Personally I think this a fairly elegant solution because the strings are modified only once, instead of every time the shape names are read and it is a very performant solution.

indiv0 commented:
Currently we deserialize `timestamp` into `f64`, which is incorrect. I will have to fix this.

We can either deserialize into `string` and leave it to the end user to parse it or we can parse it into an `chrono` `DateTime<FixedOffset>` from an `RFC3339` timestamp.

@matthewkmayer, @jimmycuadra, which do you prefer?

jimmycuadra commented:
Use chrono. Types are good!

indiv0 commented:
I tried to do it, but `chrono::DateTime<FixedOffset>` doesn't satisfy the `Default` trait so it won't work.

indiv0 commented:
Some example responses for testing purposes:

```
<ListUsersResponse xmlns=\"https://iam.amazonaws.com/doc/2010-05-08/\">
    <ListUsersResult>
        <IsTruncated>false</IsTruncated>
        <Users>
            <member>
                <Path>/</Path>
                <UserName>rusoto-testing</UserName>
                <Arn>arn:aws:iam::577584039238:user/rusoto-testing</Arn>
                <UserId>AIDAJQ4633LECA2IHI5SG</UserId>
                <CreateDate>2016-04-12T05:46:24Z</CreateDate>
            </member>
        </Users>
    </ListUsersResult>
    <ResponseMetadata>
        <RequestId>b8d32c88-04a9-11e6-a48b-8f9d6fb677f2</RequestId>
    </ResponseMetadata>
</ListUsersResponse>
```

```
<?xml version=\"1.0\"?>
<CreateQueueResponse xmlns=\"http://queue.amazonaws.com/doc/2012-11-05/\">
    <CreateQueueResult>
        <QueueUrl>https://sqs.us-east-1.amazonaws.com/577584039238/test-queue</QueueUrl>
    </CreateQueueResult>
    <ResponseMetadata>
        <RequestId>8dfe734b-fccd-5c8a-87da-3cc65d55e41e</RequestId>
    </ResponseMetadata>
</CreateQueueResponse>
```

matthewkmayer commented:
When done this will close #224 .

indiv0 commented:
I've submitted an issue to chrono regarding [implementing the `Default` trait for `DateTime<FixedOffset>`](https://github.com/lifthrasiir/rust-chrono/issues/73). If it gets accepted, then we'll be able to deserialized timestamps to that type directly. If not, we'll just stick with deserializing to `String`.

indiv0 commented:
@matthewkmayer, @jimmycuadra so I need a bit of help.

IAM is wierd in that it needs flattening of responses like SQS does (as you can see from the example responses above). @matthewkmayer managed to fix this for SQS deserialization by looking for the [`flattened`](https://github.com/rusoto/rusoto/blob/master/codegen/src/botocore.in.rs#L142) `bool` for the `Member`s.

This fix does not work for `IAM`, since `IAM` does not include a `flattened` field for members, but instead for some reason specifies a [`resultWrapper`](https://github.com/boto/botocore/blob/master/botocore/data/iam/2010-05-08/service-2.json#L1446) field on the operation [`outputs`](https://github.com/rusoto/rusoto/blob/master/codegen/src/botocore.in.rs#L116).

Now, this is problematic because `outputs` are used when generating the actual operation methods, but I need to know the `resultWrapper` when generating the deserializers, which are done exclusively via the `Member` shapes.

So my question is this: how can I find out the `resultWrapper` for an operation when generating a deserializer?

This question also leaves me unsatisfied because this doesn't seem like quite the right solution, since we're association operations with deserializers, which could lead to issues (what if there's an operation which does wrap the result, but another doesn't?).

Thoughts, ideas?

indiv0 commented:
Actually I came up with an alternative.

I could generate the deserializers serializers for `ListUsersResult` as well, then in the `list_users` operation method I could deserialize that from the XML, then simply do something like this:

```rust
let result: ListUsersResult = some_deserializer_function(stack);
let real_result = ListUsersResponse {
    users: result.users,
    ...
};
```

jimmycuadra commented:
It might be helpful to see how boto handles this.

indiv0 commented:
@jimmycuadra, @matthewkmayer: per the [issue](https://github.com/lifthrasiir/rust-chrono/issues/73) I've submitted to `rust-chrono`, @lifthrasiir suggested using `Option<DateTime<FixedOffset>>` as the type for the `timestamp`, instead of `String`, as we [currently do](https://github.com/rusoto/rusoto/blob/fc8a408d34cc1eff1d1ee2971d7d760fc16de602/codegen/src/generator/mod.rs#L99).

Is this viable? I haven't actually tested it yet, but an educated guess tells me that it will work in the general case but for queries which optionally return timestamps we'll end up with `Option<Option<DateTime<FixedOffset>>>` with the way deserializers are currently generated.

Let me know if I'm wrong and this solution will work (that'd be great!).

matthewkmayer commented:
@adimarco can you pipe in with comments on thoughts above?

adimarco commented:
Well, I'm all for types.  I'm also all for taking a step back and looking at the broader picture.  

`Default::default()` was used in the codegen at the outset primarily out of laziness, and because it was less generated and more hand-written at that point (I'll take the blame for that).  

If the field isn't actually optional, we shouldn't wrap it in an Option.  If it's supposed to have a value and one isn't provided when it's parsed, that should be handled as an error because it is one.  It's also a lot cleaner than trying to contrive a default value for the data type.  You could argue for a long integer timestamp set to 0, but that's really just a bandaid.

Which is all a roundabout way of saying: I think the correct way to do this is to stop using `Default::default()` in the codegen, since it's machine generated anyway, and there's no excuse for generated code not being meticulously precise.

So - I will open an issue for that.  I'm neck deep in the codegen right now anyway implementing typed errors and unit tests everywhere (massive new PR coming for that as soon as #261 gets merged), so it's actually a pretty easy thing to do while I'm in there.  It dovetails nicely with that work.

adimarco commented:
Opened #272 for this.  A quick peek says it's only the `query` protocol codegen (i.e., SQS) that's going to need a serious overhaul, and it could probably use one anyway.

indiv0 commented:
@adimarco, great. I'll put the `DateTime` stuff on hold for now and work on trying to fix the response flattening instead.

indiv0 commented:
I've rebased this and I am continuing work on it.

matthewkmayer commented:
Is this ready for review?

indiv0 commented:
Unfortunately not yet. I'm still working on solving the issue I last discussed.

indiv0 commented:
Per Jimmy Cuadra's advice, I attempted to understand how Botocore handles the issue I described [here](https://github.com/rusoto/rusoto/pull/235#issuecomment-213615345).

I seem to be missing a key piece of information, because it appears that Botocore adds `resultWrapper` to a `serialization` dict [here](https://github.com/boto/botocore/blob/aecb0d17616d9ca213a64411cefbc6242c2bee54/botocore/model.py#L89-L118). Then, when parsing a shape from XML, it uses this information to find the correct node [here](https://github.com/boto/botocore/blob/7184322a06f7b50f78eb2bfb618b6ef9b44d4b89/botocore/parsers.py#L455-L458). However, this leaves the question of how Botocore maps an `output` shape to a wrapping shape, unless I'm missing something.

Could anyone take a look and clarify?

cmsd2 commented:
Ah, so on further inspection of the rust code, the deserialisation methods take the expected tag name.
I've used this to pass in the name of the wrapper element if present.

Wasn't sure where to leave this so I put it over on your fork:
https://github.com/indiv0/rusoto/pull/1

indiv0 commented:
I've only skimmed your PR for now, but it looks like that will resolve the main blocker for this issue! Thank you for that :)

I should be able to finish up the rest of this PR pretty soon.

indiv0 commented:
@cmsd2 so I've given your PR a try (and added to my PR locally as I believe it's the right step), but it still gives errors while parsing certain IAM responses (e.g. `list_users`).

I'll keep working on it locally to try to fix it.

I've uploaded a simple test case for it and merged your changes to my local branch, if you want to give it a try as well.

indiv0 commented:
So it looks like the issue is because `ListUsersResponse` follows the format:

```rust
<ListUsersResponse xmlns=\"https://iam.amazonaws.com/doc/2010-05-08/\">
    <ListUsersResult>
        ...
        <Users>
            <member>
                <Path>/</Path>
                ...
            </member>
        </Users>
    </ListUsersResult>
    ...
</ListUsersResponse>
```

while our deserialization code expects the following:

```rust
<ListUsersResponse xmlns=\"https://iam.amazonaws.com/doc/2010-05-08/\">
    <ListUsersResult>
        ...
        <Users>
            <User>
                <Path>/</Path>
                ...
            </User>
        </Users>
    </ListUsersResult>
    ...
</ListUsersResponse>
```

(i.e. `member` is used for the instance tag name instead of `User`).

My current inquiry into this problem is the fact that there doesn't appear to be any indication for when this type of response is returned (i.e. akin to `flattened` or `resultWrapper`) which leads me to believe this occurs for all `query` service responses which return a list of structs. I'll perform some local tests to verify this.

@matthewkmayer @cmsd2 any suggestions for fixing this?

cmsd2 commented:
I've had a look through the api documentation for iam.

All of the List* actions have a member element in them.
In addition, these actions also have member elements in:

GetAccountAuthorizationDetails
GetAccountSummary (actuall has an entry element, not a member element)
GetContextKeysForCustomPolicy
GetContextKeysForPrincipalPolicy
GetGroup
GetInstanceProfile
GetOpenIDConnectProvider
SimulateCustomPolicy
SimulatePrincipalPolicy

cmsd2 commented:
more testing required, but it seems to work for iam's ListUsers:

https://github.com/indiv0/rusoto/pull/2

matthewkmayer commented:
Looks like this service uses the botocore definition of a `shape` having `members` more literally than the other services, where the field us usually specified by a name identifier.

While I don't have a great idea on how to fix this particular service, it's highlighting how we need to use more of the sample responses available to us in the botocore git submodule.  Then we can experiment with various methods of correctly parsing things.  https://github.com/boto/botocore/blob/develop/tests/unit/response_parsing/xml/responses/iam-list-users.xml has a sample response for list-users, for example.

indiv0 commented:
Looks like your PR breaks at least `SqsClient::list_queues`. Looking into it now.

indiv0 commented:
The problem is because for simple lists, the result comes in the form:

```xml
<ListQueuesResponse>
    <ListQueuesResult>
        <QueueUrl>...</QueueUrl>
        <QueueUrl>...</QueueUrl>
    </ListQueuesResult>
</ListQueuesResponse>
```

and not:

```xml
<ListQueuesResponse>
    <ListQueuesResult>
        <QueueUrls>
            <QueueUrl>...</QueueUrl>
            <QueueUrl>...</QueueUrl>
        </QueueUrls>
    </ListQueuesResult>
</ListQueuesResponse>
```

cmsd2 commented:
That particular result is actually a flattened list according to botocore data. That should be enough to distinguish between the two, I just didn't take account of that.

Agreed on making better use of test data. The integration tests we have at the moment are not making it easy

cmsd2 commented:
updated the pr to avoid stepping in to that non-existant QueueUrls element by checking if the list is flattened. cleaned up the logic a tiny bit to make it less impossible to read.

adimarco commented:
I just yanked this out of a much bigger PR that hasn't been merged yet.  Looks like it would be a lot of use here.  #368 allows the whole HTTP request/response to be mocked for unit tests.  You could very easily test how the generated code deals with arbitrary responses, and write some proper tests for it.

cmsd2 commented:
I started testing a version of the iam code rebased against the new autogen'd tests.

cargo test --features "sqs iam" --lib

3 failures:
iam::protocol_tests::test_parse_iam_list_access_keys
iam::protocol_tests::test_parse_iam_list_instance_profiles
iam::protocol_tests::test_parse_iam_list_signing_certificates

the changes i had to make were to step into non-flattened lists and to use "member" as the list location name. that's about it.

these failures look like maps possibly with some non-standard weirdness involved.

matthewkmayer commented:
What can we do to finish up the IAM work?

cmsd2 commented:
ListInstanceProfiles:
the InstanceProfiles in the list seem to have an Id property instead of the expected InstanceProfileId
how should I special-case a field-rename?

ListSigningCertificates:
The ListSigningCertificatesResult contains an extra unexpected UserName element that isn't in the shape.

ListAccessKeys:
The ListAccessKeysResult also contains an extra unexpected UserName element that isn't in the shape.

The aws api docs have the same discrepancy between sample xml and documented response shapes.
I don't know what the correct fix is.

matthewkmayer commented:
Digging into `ListInstanceProfiles` I think that's the same issue I'm running into in https://github.com/rusoto/rusoto/pull/375 .  Shapes wrap shapes and you need to go down to the primitive shape to extract the required data.

I also see unexpected elements in EC2 sample responses, but I haven't figured out what to do about them yet.

cmsd2 commented:
case id: 1874502911
"""
I've confirmed that these examples do not reflect the actual API responses; there are no circumstances where we should expect these undocumented elements. I'll work with our documentation team to correct these examples.
"""

matthewkmayer commented:
What can we do to help finish this off?  Are there botocore issues we can weigh in on/make PRs for?  Did my response above help at all? 😄

cmsd2 commented:
er right, yes. sorry been busy with an ed-x course.

the example responses that botocore has are precise copies of the examples on the aws documentation site, so they copied mistakes and all. it would be good if botocore could fix their response files.
we could temporarily ignore, or maybe better mask their files with our own folder of response files.

i have no idea what else is required after fixing these short term problems.

@indiv0 comments?

cmsd2 commented:
I've masked the botocore response files for those 3 iam calls with some hand-modified copies which I think match the spec.

https://github.com/cmsd2/rusoto/tree/iam-patched-responses

indiv0 commented:
Hello, sorry for the lack of replies, I've been a bit busy lately.

@cmsd2 I'd be OK with masking the incorrect responses until botocore fixes them (we should probably open an upstream issue for this, though).

As for what needs to be done, assuming that @cmsd2's commits fix XML deserialization for IAM, then the remaining tasks are:

* [ ] Verify serialization was not broken for other services (e.g. SQS)
* [ ] Fix the way the generator generates function names
    * Currently it attempts to naively snake case the type, which doesn't work for types like `ListSAMLProviders`, which becomes `list_s_a_m_l_providers`
* [ ] Add tests
    * ..?
* [ ] Update `README.md`
* [ ] Squash commits
* [ ] Wait for https://github.com/rusoto/rusoto/pull/386 to land, then rebase the code to utilize it's `capitalize_first_letter_of_string` function instead.
* [ ] Rebase on master
* ..?

cmsd2 commented:
I think https://github.com/whatisinternet/inflector/commit/6b5614c721ece302f4e98d4f370563f184672bc3 addresses the snake case item on the list.

indiv0 commented:
@cmsd2 I've added your commits to the PR branch, with a few modifications.

Of particular note, I modified the testgen code to generate tests for methods which don't take any params (e.g. `IamClient::get_account_summary`, which doesn't take `&request`). All of the added tests pass, except for `get_account_summary`, which gives:

```
---- iam::protocol_tests::test_parse_iam_get_account_summary stdout ----
	thread 'iam::protocol_tests::test_parse_iam_get_account_summary' panicked at 'parse error: Err(Unknown("START Expected SummaryKeyType got entry"))', /home/indiv0/src/rust/rusoto/target/debug/build/rusoto-76b9765c775ff26b/out/iam.rs:32292
stack backtrace:
   1:     0x55c0cda4333f - std::sys::backtrace::tracing::imp::write::h22f199c1dbb72ba2
   2:     0x55c0cda477cd - std::panicking::default_hook::{{closure}}::h9a389c462b6a22dd
   3:     0x55c0cda4650e - std::panicking::default_hook::h852b4223c1c00c59
   4:     0x55c0cda46bf8 - std::panicking::rust_panic_with_hook::hcd9d05f53fa0dafc
   5:     0x55c0cda46a92 - std::panicking::begin_panic::hf6c488cee66e7f17
   6:     0x55c0cda469d0 - std::panicking::begin_panic_fmt::hb0a7126ee57cdd27
   7:     0x55c0cd76e1c1 - rusoto::iam::protocol_tests::test_parse_iam_get_account_summary::hd72bba0a24a524d1
                        at /home/indiv0/src/rust/rusoto/src/iam.rs:5
   8:     0x55c0cd7b8f1b - <F as alloc::boxed::FnBox<A>>::call_box::hdb02e989a089cf4e
   9:     0x55c0cd7b2d33 - std::panicking::try::do_call::hb5542e7f62c71bb7
  10:     0x55c0cda4f296 - __rust_maybe_catch_panic
  11:     0x55c0cd7b8d89 - <F as alloc::boxed::FnBox<A>>::call_box::hb415bac963aa88f8
  12:     0x55c0cda453d0 - std::sys::thread::Thread::new::thread_start::h50b05608a499d2b2
  13:     0x7f5f9ed23453 - start_thread
  14:     0x7f5f9e84f7de - __GI___clone
  15:                0x0 - <unknown>
```

More parser errors, yay! I'm looking into this and working on the remainder of the PR ATM.

indiv0 commented on codegen/src/generator/query.rs:
```diff
@@ -279,41 +313,63 @@ fn generate_struct_deserializer(name: &str, shape: &Shape, service: &Service) ->
-            "\"{location_name}\" => {{
+            "
+            // {child_shape:?}
+            // {member:?}
```
We probably want to remove these lines prior to merging the PR, or at least make them debug-mode only. Otherwise, this will greatly inflate our memory usage during codegen I think (which is already a problem for us).

matthewkmayer replied:
I agree: let's pull those lines out.

indiv0 replied:
Alright will do.

indiv0 commented:
Alright, so https://github.com/rusoto/rusoto/pull/235/commits/1cb85eb4a47b5c914f548012292899a95f5a556b should fix the last known XML deserialization issue (handling non-flattened maps). Currently, all SQS and IAM tests pass successfully.

Other than rebasing/commit cleanup (which I'm working on now), the only two things remaining are to write tests and decide what to do in regards to [this](https://github.com/rusoto/rusoto/pull/235#discussion_r81618522).

@jimmycuadra @matthewkmayer @adimarco should I write a full unit test suite for every IAM function or are we satisfied with the generated tests?

matthewkmayer commented:
I'm fine with tests that just check our parsing of sample responses from botocore.  A single integration test that lists users or roles might be helpful for demonstration purposes but isn't a requirement.

Thanks for everyone's hard work on this!

indiv0 commented:
Alright well since we already have `get_user` and `list_users` tests for IAM, I'll leave them as-is. @matthewkmayer thoughts on [this](https://github.com/rusoto/rusoto/pull/235#discussion_r81618522)?

indiv0 commented:
Alright, well, I think that about wraps it up! I've run the integration tests with all features enabled locally (compilation used 5043 MB of RAM!!), they were successful (except S3, I couldn't get it configured right). Barring any crazy results from Travis CI, this PR is good to go!

I've rebased it on master and squashed the commits down to a more logical set, and fixed up the commit logs.

I am now formally requesting a code review of the PR, unless anyone has any last minute changes/suggestions.

Thank you very much to everyone who contributed on this PR, and I'd like to give an extra special thank you to @cmsd2 for his help!

matthewkmayer commented:
Appveyor killed our job since it took more than an hour to run:

`Build execution time has reached the maximum allowed time for your plan (60 minutes).`

I don't think this blocks review of the ticket, but we will need to make another one for specifically addressing IAM compile speeds.  Our compile times on Travis and Appveyor will only get worse as we bring S3 into the codegen world.

matthewkmayer commented:
Apples to oranges due to lots of codegen changes and different versions of `rustc` but here's the lines of code this branch generates for LLVM:

`cargo rustc --features "all" -- --emit llvm-ir && wc -l target/debug/rusoto.ll`

`5394108 target/debug/rusoto.ll`

matthewkmayer commented:
I restarted the appveyor build and it squeaked in at 58 minutes of build time.  Definitely need work to improve that.

matthewkmayer commented on tests/sqs.rs:
```diff
@@ -0,0 +1,20 @@
+#![cfg(feature = "iam")]
```
Feature should be `sqs` instead of `iam`.

indiv0 replied:
Noted, fixing now.

matthewkmayer commented on codegen/src/serialization.rs:
```diff
@@ -0,0 +1,87 @@
+use std::collections::BTreeMap;
```
Is the purpose of this whole file to provide Serde with an override so it knows how to handle shape names that need to be deserialized with `capitalize_first` called on them?

indiv0 replied:
Yes. Silly, I know.

indiv0 replied:
Would you like me to add a comment to the top of the file about this, just so it's clear in the future?

matthewkmayer replied:
Please.  It's a bit of an edge case we don't run into with other uses of Serde in the project so documenting it will help us remember why it's there.

matthewkmayer commented on codegen/tests/unit/responses/iam-list-access-keys.xml:
```diff
@@ -0,0 +1,20 @@
+<ListAccessKeysResponse xmlns="https://iam.amazonaws.com/doc/2010-05-08/">
```
This file is in the codegen botocore's dir - can we refer to that one?  I assume this file is left over from testing.

indiv0 replied:
This and the other files were added because (per the discussion earlier in the PR), the botocore versions of the responses are broken, so we'll use our own corrected ones until they're fixed upstream.

matthewkmayer commented on codegen/tests/unit/responses/iam-list-instance-profiles.xml:
```diff
@@ -0,0 +1,26 @@
+<ListInstanceProfilesResponse xmlns="https://iam.amazonaws.com/doc/2010-05-08/">
```
Another sample file we could snag from botocore.

matthewkmayer commented on codegen/tests/unit/responses/iam-list-signing-certificates.xml:
```diff
@@ -0,0 +1,31 @@
+<ListSigningCertificatesResponse xmlns="https://iam.amazonaws.com/doc/2010-05-08/">
```
Another sample file we could snag from botocore.

matthewkmayer commented:
Definitely worth doing a release when this lands!

matthewkmayer commented:
@cmsd2 have you received any more info on the AWS support case for the sample responses?  Perhaps we should raise an issue upstream with [botocore](https://github.com/boto/botocore).  The [sample files for IAM](https://github.com/boto/botocore/tree/develop/tests/unit/response_parsing/xml/responses) haven't been updated for two to four years.

Not blocking this PR but something we should keep track of.

indiv0 commented:
@matthewkmayer per your review, I've changed the following:

* Fixed the feature check to look for `iam` instead of `sqs`
* Left the custom botocore response examples as-is, per my comment regarding them
* Fixed the testgen to generate tests for actions which we can't case properly
    * Specifically, tests for actions like `ListMFADevices` weren't being generated because `iam-list-mfa-devices.xml` was incorrectly being determined to be the action `ListMfaDevices` (improper casing). Since we can't determine the proper casing from the filename in cases like that (short of adding explicit exceptions during testgen), I've modified the testgen for both the query and EC2 generators to attempt to find matching actions case-insensitively. This is done by a [wrapper](https://github.com/rusoto/rusoto/pull/235/files#diff-c5111460d6c81675788a12b76e26e676R6) over `BTreeMap<String, V>::get`.
* You mentioned that testgen didn't generate a test for `ListSigningCertificates`, but it works on my end

That should cover all of the notes you made during the review. Let me know if you have any further requests!

indiv0 commented:
Actually, I missed [your comment](https://github.com/rusoto/rusoto/pull/235#discussion_r81897175) regarding my suggestion. I'll put that in right now.

indiv0 commented:
There we go, https://github.com/rusoto/rusoto/pull/235/commits/fa76d5112cb6f9a97a8220de1b1f6464268e8fc7 adds the explanatory comment.

Ready for review again :)

cmsd2 commented:
@matthewkmayer 
I'm not expecting any more back from AWS. They may eventually update their documentation.

matthewkmayer commented:
Thanks @cmsd2 .  

@indiv0 `ListSigningCertificates` missing may be a copy/paste error on my part - I see it now.

This PR looks good to go! 🎉 Thanks for everyone's hard work.

indiv0 commented:
@matthewkmayer do we want to retry the Appveyor build? If not, feel free to merge this in :)
